app/src/main/java/nodomain/freeyourgadget/gadgetbridge/export/GPXExporter.java

```diff
@@ -117,29 +119,53 @@ private boolean exportTrackPoint(XmlSerializer ser, ActivityPoint point, String
         if (description != null) {
             ser.startTag(NS_DEFAULT, "desc").text(description).endTag(NS_DEFAULT, "desc");
         }
-        ser.startTag(NS_DEFAULT, "src").text(source).endTag(NS_DEFAULT, "src");
+        //ser.startTag(NS_DEFAULT, "src").text(source).endTag(NS_DEFAULT, "src");
 
-        exportTrackpointExtensions(ser, point);
+        exportTrackpointExtensions(ser, point, trackPoints);
 
         ser.endTag(NS_DEFAULT, "trkpt");
 
         return true;
     }
 
-    private void exportTrackpointExtensions(XmlSerializer ser, ActivityPoint point) throws IOException {
+    private void exportTrackpointExtensions(XmlSerializer ser, ActivityPoint point, List<ActivityPoint> trackPoints) throws IOException {
         if (!includeHeartRate) {
             return;
         }
 
         int hr = point.getHeartRate();
         if (!HeartRateUtils.isValidHeartRateValue(hr)) {
+            if(!includeHeartRateByTime) { return; }
+
+            Date time = point.getTime();
```

Comment:
I'll try to factor this out into a separate method. Unfortunately it's still O(n^2), but we can fix this later.